api/renewer.go

```diff
@@ -0,0 +1,243 @@
+// 		default:
+// 		}
+// 	}
+//
+//
+// The `DoneCh` will return if renewal fails or if the remaining lease duration
+// after a renewal is less than or equal to the grace (in number of seconds). In
+// both cases, the caller should attempt a re-read of the secret. Clients should
+// check the return value of the channel to see if renewal was successful.
+type Renewer struct {
+	sync.Mutex
+
+	client *Client
+	secret *Secret
+	grace  int
+	doneCh chan error
+	tickCh chan struct{}
+
+	stopped bool
+	stopCh  chan struct{}
+}
+
+var (
+	ErrRenewerMissingInput  = errors.New("missing input to renewer")
+	ErrRenewerMissingSecret = errors.New("missing secret to renew")
+	ErrRenewerNotRenewable  = errors.New("secret is not renewable")
+	ErrRenewerNoSecretData  = errors.New("returned empty secret data")
+
+	// DefaultRenewerGrace is the default grace period
+	DefaultRenewerGrace = 15
+)
+
+// NewRenewer creates a new renewer from the given input.
+func (c *Client) NewRenewer(i *RenewerInput) (*Renewer, error) {
+	if i == nil {
+		return nil, ErrRenewerMissingInput
+	}
+
+	secret := i.Secret
+	if secret == nil {
+		return nil, ErrRenewerMissingSecret
+	}
+
+	grace := i.Grace
+	if grace == 0 {
+		grace = DefaultRenewerGrace
+	}
+
+	return &Renewer{
+		client: c,
+		secret: secret,
+		grace:  grace,
+		doneCh: make(chan error, 1),
+		tickCh: make(chan struct{}, 5),
+
+		stopped: false,
+		stopCh:  make(chan struct{}, 1),
+	}, nil
+}
+
+// DoneCh returns the channel where the renewer will publish when renewal stops.
+// If there is an error, this will be an error.
+func (r *Renewer) DoneCh() <-chan error {
+	return r.doneCh
+}
+
+// TickCh is a channel that receives a message when a successful renewal takes
+// place.
+func (r *Renewer) TickCh() <-chan struct{} {
+	return r.tickCh
+}
+
+// Stop stops the renewer.
+func (r *Renewer) Stop() {
+	r.Lock()
+	if !r.stopped {
+		close(r.stopCh)
+		r.stopped = true
+	}
+	r.Unlock()
+}
+
+// Renew starts a background process for renewing this secret. When the secret
+// is has auth data, this attempts to renew the auth (token). When the secret
+// has a lease, this attempts to renew the lease.
+//
+// This function will not return if nothing is reading from doneCh (it blocks)
+// on a write to the channel.
+func (r *Renewer) Renew() {
+	if r.secret.Auth != nil {
+		r.doneCh <- r.renewAuth()
+	} else {
+		r.doneCh <- r.renewLease()
+	}
+}
+
+// renewAuth is a helper for renewing authentication.
+func (r *Renewer) renewAuth() error {
+	if !r.secret.Auth.Renewable || r.secret.Auth.ClientToken == "" {
+		return ErrRenewerNotRenewable
+	}
+
+	client, token := r.client, r.secret.Auth.ClientToken
+
+	for {
+		// Check if we are stopped.
+		select {
+		case <-r.stopCh:
+			return nil
+		default:
+		}
+
+		// Renew the auth.
+		renewal, err := client.Auth().Token().RenewSelfAsToken(token, 0)
+		if err != nil {
+			return err
+		}
+
+		// Push a message that a renewal took place.
+		select {
+		case r.tickCh <- struct{}{}:
```

Comment:
Rather than an empty struct, I think it would be useful to instead just push down the Auth info, or some way for the user to be able to introspect the new information without then having to perform a lookup on the lease or token. I haven't thought deeply about what information should be sent, but at a minimum maybe a struct with the new TTL?